packages/library-base/src/commonMain/kotlin/io/realm/kotlin/log/RealmLog.kt

```diff
@@ -0,0 +1,390 @@
+import io.realm.kotlin.internal.interop.SynchronizableObject
+import io.realm.kotlin.internal.platform.createDefaultSystemLogger
+import io.realm.kotlin.log.RealmLog.add
+import io.realm.kotlin.log.RealmLog.addDefaultSystemLogger
+
+/**
+ * Global logger class used by all Realm components.
+ *
+ * By default all logs will go to a default system logger that will depend on the system. See
+ * [addDefaultSystemLogger] for more details.
+ *
+ * Custom loggers can be added by registering a class implementing [RealmLogger] using [add].
+ *
+ * When logging messages, it it possible to use a subset of String format options as known from
+ * Java. Only `%s`, `%d` and `%f` are supported. See https://stackoverflow.com/a/64499248/1389357
+ * and https://youtrack.jetbrains.com/issue/KT-25506 for more information.
+ */
+public object RealmLog {
+
+    /**
+     * The current [LogLevel]. Changing this will affect all registered loggers.
+     */
+    public var level: LogLevel = LogLevel.INFO
+        set(value) {
+            RealmInterop.realm_set_log_level(value.toCoreLogLevel())
+            field = value
+        }
+
+    // Lock preventing multiple threads modifying the list of loggers.
+    private val loggersMutex = SynchronizableObject()
+    // Reference to the currently installed system logger (if any)
+    // `internal` until we can remove the old log API
+    internal var systemLoggerInstalled: RealmLogger? = null
+    // Kotlin Multiplatform are currently lacking primitives like CopyOnWriteArrayList. We could
+    // use `io.realm.kotlin.internal.interop.SynchronizableObject`, but it would require locking
+    // when reporting a log statement which feel a bit heavy, so instead we have added locks around
+    // all modifications to this array (which are expected to be rare) and the `doLog` method must
+    // copy this reference before using it.
+    private var loggers: MutableList<RealmLogger> = mutableListOf()
+
+    init {
+        addDefaultSystemLogger()
+        RealmInterop.realm_set_log_callback(
+            level.toCoreLogLevel(),
+            object : LogCallback {
+                override fun log(logLevel: Short, message: String?) {
+                    // Create concatenated up front, since Core should already filter messages
+                    // not within the log range.
+                    val level: LogLevel = fromCoreLogLevel(CoreLogLevel.valueFromPriority(logLevel))
+                    doLog(
+                        level,
+                        null,
+                        if (message.isNullOrBlank()) { null } else { "[Core] $message" }
+                    )
+                }
+            }
+        )
+    }
+
+    /**
+     * Add a logger that will be notified on log events that are equal to or exceed the currently
+     * configured [level].
+     *
+     * @param logger logger to add.
+     */
+    public fun add(logger: RealmLogger) {
+        loggersMutex.withLock {
+            loggers.add(logger)
```

Comment:
This doesn't make the `loggers` stable while using it. You need to make a single update of it. 
```
val copyOfOldLoggers = ...
copyOfOldLoggers.add(logger)
loggers = copyOfOldLoggers
```
or something like that. Just like you do it for `remove`.

Reply:
Good catch 👍